### PR TITLE
[RSDK-8982, RSDK-9167, RSDK-8979] - Add SetStreamOptions to stream server

### PR DESCRIPTION
**Pull request by seanavery**
## Description

[RSDK-8982](https://viam.atlassian.net/browse/RSDK-8982), [RSDK-9167](https://viam.atlassian.net/browse/RSDK-9167) [RSDK-8979](https://viam.atlassian.net/browse/RSDK-8979)

This PR adds the `SetStreamOptions` endpoint to the stream server. 
- Validates the requests and performs a hot swap on the video source with a `NewResizeVideoSource`.
- If Resolution is provided, we dispatch a `Resize` msg to the `streamState` handler.
  - If in Passthrough mode, the state transitions to gostream.
  - Prevents attempting to use passthrough during `Increment` and `Decrement` calls.
- If Resolution is not provided, we dispatch a `Reset` msg to the `streamState` handler.
  - Attempts to restart passthrough if available.
- In order to unify RTP `Sequence Number` handling between GoStream and Passthrough paths we now override the Sequence Number header field in the `WriteRTP` method.

## Tests
- CI Tests:
  - Request Validation ✅ 
  - GetImage Fake camera resize swap ✅ 
  - RTPPassthrough fake camera downgrade state transition ✅
  - Video tracks stay alive through resize transition ✅ 
  - Verifies prevention of RTPPassthrough when resized and peers Increment/Decrement ✅ 
  - Gostream back Passthrough reset state transition ✅ 
 - Manual tests (using `NewStreamServiceClient` script as seen below):
   - GetImage livestreams still work ✅ 
     - webcam ✅ 
     - fake cam ✅ 
     - viamrtsp ✅ 
   - Passthrough livestreams still work ✅
     - fake cam ✅ 
     - viamrtsp ✅ 
   - GetImage resize livestreams ✅ 
     - webcam ✅ 
     - fake cam ✅
     - viamrtsp ✅ 
   - Passthrough->GetImage resize livestream ✅
     - fake cam ✅ 
     - viamrtsp ✅ 
   - GetImage->Passthrough reset livestream ✅ 
     - fake cam ✅ 
     - viamrtsp ✅ 
     
___
### Demos
- viamrtsp passthrough -> resize getimage stream via SetStreamOptions

https://github.com/user-attachments/assets/0137c2b2-2b17-49cb-a7c6-637e9e8c3813

- passthrough <-> gostream getimage loop using resize and reset

https://github.com/user-attachments/assets/f6104227-0376-4a26-bfce-c9fdf0e180e4

___
### TODO:
- [x] Test that video tracks survive SetSteamOptions swap.
- [x] Make sure we are safe wrt locking strategy.
- [x] Resize state transition from passthrough -> get_image.
- [x] Figure out how to test stream state transition for passthrough -> get_image.
- [x] Figure out how to prevent passthrough attempts when in resized mode.
- [x] Make sure resizes are working with viamrtsp camera source.
- [x] Fix flaky handoff between Gostream and Passthrough RTP sources.

[RSDK-8982]: https://viam.atlassian.net/browse/RSDK-8982?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RSDK-9167]: https://viam.atlassian.net/browse/RSDK-9167?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RSDK-8979]: https://viam.atlassian.net/browse/RSDK-8979?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

___
### Test Script

```golang
package main

import (
	"context"
	"time"

	streampb "go.viam.com/api/stream/v1"
	rgrpc "go.viam.com/rdk/grpc"
	rdklogger "go.viam.com/rdk/logging"
	"go.viam.com/utils/rpc"
)

func main() {
	logger := rdklogger.NewLogger("main")
	addr := "..."
	keyID := "..."
	apiKey := "..."
	camName := "..."
	ctx := context.Background()
	conn, err := rgrpc.Dial(ctx, addr, logger, rpc.WithEntityCredentials(
		keyID,
		rpc.Credentials{
			Type:    rpc.CredentialsTypeAPIKey,
			Payload: apiKey,
		}))
	if err != nil {
		logger.Error("Failed to dial: %v", err)
		return
	}
	livestreamClient := streampb.NewStreamServiceClient(conn)
	if livestreamClient == nil {
		logger.Error("Failed to create stream service client")
		return
	}
	opts, err := livestreamClient.GetStreamOptions(ctx, &streampb.GetStreamOptionsRequest{
		Name: camName,
	})
	if err != nil {
		logger.Error("Failed to get stream options: %v", err)
		return
	}
	logger.Info("Stream options: %v", opts)
	for {
		for _, resolution := range opts.Resolutions {
			logger.Info("Resizing to %v", resolution)
			_, err = livestreamClient.SetStreamOptions(ctx, &streampb.SetStreamOptionsRequest{
				Name:       camName,
				Resolution: resolution,
			})
			if err != nil {
				logger.Error("Failed to set stream options: %v", err)
				return
			}
			time.Sleep(5 * time.Second)
			logger.Info("Resetting to default")
			_, err = livestreamClient.SetStreamOptions(ctx, &streampb.SetStreamOptionsRequest{
				Name: camName,
			})
			if err != nil {
				logger.Error("Failed to set stream options: %v", err)
				return
			}
			time.Sleep(5 * time.Second)
		}
	}
}

```

### CODE REVIEW

**Review by hexbabe — APPROVED** [review]
Stream state and server are still maturing concepts in my head. Tried my best to focus on code quality stuff and learned stuff about the former along the way :)

File: robot/web/stream/state/state.go [hunk 1]
```diff
@@ -61,6 +64,7 @@ func New(
 		robot:     r,
 		msgChan:   make(chan msg),
 		tickChan:  make(chan struct{}),
+		resized:   false,
```
[the remark this answers is not in the document]
Comment (seanavery) [reply robot/web/stream/state/state.go]: Agreed. Changed to `isResized`.

File: robot/web/stream/server.go [hunk 2]
```diff
@@ -372,6 +374,55 @@ func (server *Server) GetStreamOptions(
...
+// to the given width and height.
+func (server *Server) SetStreamOptions(
+	ctx context.Context,
+	req *streampb.SetStreamOptionsRequest,
+) (*streampb.SetStreamOptionsResponse, error) {
+	if req.Name == "" {
+		return nil, errors.New("stream name is required")
+	}
+	if req.Resolution == nil {
+		return nil, errors.New("resolution is required")
```
[the remark this answers is not in the document]
Comment (seanavery) [reply robot/web/stream/server.go]: Good call. Added name to error returns and more details where appropriate.

File: robot/web/stream/stream_test.go [hunk 3]
```diff
@@ -356,3 +356,134 @@ func TestGetStreamOptions(t *testing.T) {
...
+	test.That(t, len(listResp.Names), test.ShouldEqual, 2)
+
+	getStreamOptionsResp, err := livestreamClient.GetStreamOptions(ctx, &streampb.GetStreamOptionsRequest{
+		Name: "fake-cam-0-0",
+	})
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, getStreamOptionsResp, test.ShouldNotBeNil)
+	test.That(t, len(getStreamOptionsResp.Resolutions), test.ShouldEqual, 5)
+
+	// Test setting stream options with invalid name
```
[the remark this answers is not in the document]
Comment (seanavery) [reply robot/web/stream/stream_test.go]: Yeah good call better debugging if things fail. t.Run's should still be sequential if under a parent test fn.

File: robot/web/stream/state/state_test.go [hunk 4]
```diff
@@ -762,4 +762,133 @@ func TestStreamState(t *testing.T) {
...
+		test.That(t, s.Increment(), test.ShouldBeNil)
+		testutils.WaitForAssertion(t, func(tb testing.TB) {
+			test.That(tb, subscribeRTPCount.Load(), test.ShouldEqual, 1)
+			test.That(tb, unsubscribeCount.Load(), test.ShouldEqual, 0)
+		})
+		// WriteRTP is called
+		<-writeRTPCalledCtx.Done()
+
+		// Call Resize to simulate a resize event. This should stop rtp_passthrough and start gostream.
+		logger.Info("calling Resize() should stop rtp_passthrough and start gostream")
```
[the remark this answers is not in the document]
Comment (seanavery) [reply robot/web/stream/state/state_test.go]: Yep should be ok to nest t.Run tests here.

File: robot/web/stream/stream_test.go [hunk 5]
```diff
@@ -356,3 +356,134 @@ func TestGetStreamOptions(t *testing.T) {
...
+	})
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, res, test.ShouldNotBeNil)
+	logger.Info(conn.PeerConn().CurrentLocalDescription().SDP)
+	testutils.WaitForAssertion(t, func(tb testing.TB) {
+		videoCnt = strings.Count(conn.PeerConn().CurrentLocalDescription().SDP, "m=video")
+		test.That(tb, videoCnt, test.ShouldEqual, 2)
+	})
+
+	// Try setting stream options with RTPPassthrough enabled
```
[the remark this answers is not in the document]
Comment (seanavery) [reply robot/web/stream/stream_test.go]: Testing this specifically in `state_test.go`. This is just testing that the API call succeeds and track is still available.

File: robot/web/stream/server.go [hunk 6]
```diff
@@ -336,6 +336,8 @@ func (server *Server) GetStreamOptions(
 	ctx context.Context,
 	req *streampb.GetStreamOptionsRequest,
 ) (*streampb.GetStreamOptionsResponse, error) {
+	server.mu.RLock()
```
Comment (nicksanford) [inline robot/web/stream/server.go:339]: Why do we need to add this read lock?

File: robot/web/stream/server.go [hunk 7]
```diff
@@ -372,6 +374,58 @@ func (server *Server) GetStreamOptions(
...
+	}
+	if req.Resolution == nil {
+		return nil, fmt.Errorf("resolution is required to resize stream %q", req.Name)
+	}
+	if req.Resolution.Width <= 0 || req.Resolution.Height <= 0 {
+		return nil, fmt.Errorf(
+			"invalid resolution to resize stream %q: width (%d) and height (%d) must be greater than 0",
+			req.Name, req.Resolution.Width, req.Resolution.Height,
+		)
+	}
```
Comment (nicksanford) [inline robot/web/stream/server.go:394]: Lets move this into a validation function

File: robot/web/stream/server.go [hunk 8]
```diff
@@ -372,6 +374,58 @@ func (server *Server) GetStreamOptions(
...
+		return nil, fmt.Errorf("failed to resize video source for stream %q: %w", req.Name, err)
+	}
+	streamState, ok := server.nameToStreamState[req.Name]
+	if !ok {
+		return nil, fmt.Errorf("stream state not found with name %q", req.Name)
+	}
+	err = streamState.Resize()
+	if err != nil {
+		return nil, fmt.Errorf("failed to resize stream %q: %w", req.Name, err)
+	}
```
Comment (nicksanford) [inline robot/web/stream/server.go:408]: Could we move this logic into `resizeVideoSource`?

File: robot/web/stream/state/state.go [hunk 9]
```diff
@@ -185,6 +200,10 @@ func (state *StreamState) sourceEventHandler() {
 			if state.activeClients == 0 {
 				state.tick()
 			}
+		case msgTypeResize:
+			state.logger.Debug("resize event received")
+			state.isResized = true
```
Comment (nicksanford) [inline robot/web/stream/state/state.go:205]: Can we ever go back to not resized?

File: robot/web/stream/state/state.go [hunk 10]
```diff
@@ -271,7 +302,7 @@ func (state *StreamState) tick() {
 	case state.streamSource == streamSourcePassthrough:
 		// no op if we are using passthrough & are healthy
 		state.logger.Debug("still healthy and using h264 passthrough")
-	case state.streamSource == streamSourceGoStream:
+	case state.streamSource == streamSourceGoStream && !state.isResized:
```
Comment (randhid) [inline robot/web/stream/state/state.go:305]: What does a user do to hit this state after they've selected a resolution in the UI?

File: robot/web/stream/server.go (hunk 6 above)
Comment (nicksanford) [reply robot/web/stream/server.go:339]: The mutex is protecting these struct members:
```
	nameToStreamState       map[string]*state.StreamState
	activePeerStreams       map[*webrtc.PeerConnection]map[string]*peerState
	activeBackgroundWorkers sync.WaitGroup
	isAlive                 bool

```

I don't see them being accessed so I don't understand why this mutex is being taken

Comment (seanavery) [reply robot/web/stream/server.go]: Good call - I am just pulling the camera from robot not accessing any of these. Removed.

File: robot/web/stream/server.go (hunk 7 above)
Comment (seanavery) [reply robot/web/stream/server.go]: Sounds good - moved to helper.

File: robot/web/stream/server.go (hunk 8 above)
Comment (seanavery) [reply robot/web/stream/server.go]: Yep -- moved all resize logic into `resizeVideoSource`.

File: robot/web/stream/state/state_test.go [hunk 11]
```diff
@@ -762,4 +762,133 @@ func TestStreamState(t *testing.T) {
 			test.That(tb, stopCount.Load(), test.ShouldEqual, 3)
 		})
 	})
+
+	t.Run("when in rtppassthrough mode and a resize occurs test downgrade path to gostream", func(t *testing.T) {
+		var startCount atomic.Int64
```
[the remark this answers is not in the document]
Comment (seanavery) [reply robot/web/stream/state/state_test.go:767]: `go test -timeout 30s -race -failfast -run ^TestSetStreamOptions$ go.viam.com/rdk/robot/web/stream`
ok      go.viam.com/rdk/robot/web/stream        3.255s
___
`go test -timeout 30s -race -failfast -run ^TestStreamState$/^when_in_rtppassthrough_mode_and_a_resize_occurs_test_downgrade_path_to_gostream$ go.viam.com/rdk/robot/web/stream/state`
ok      go.viam.com/rdk/robot/web/stream/state  3.056s

File: robot/web/stream/server.go [hunk 12]
```diff
@@ -336,6 +336,8 @@ func (server *Server) GetStreamOptions(
 	ctx context.Context,
```
[the remark this answers is not in the document]
Comment (seanavery) [reply robot/web/stream/server.go]: Thanks, added to PR title & desc.

File: robot/web/stream/server.go [hunk 13]
```diff
@@ -677,3 +761,21 @@ retryLoop:
...
+	}
+	if req.Resolution.Width <= 0 || req.Resolution.Height <= 0 {
+		return optionsCommandUnknown,
+			fmt.Errorf(
+				"invalid resolution to resize stream %q: width (%d) and height (%d) must be greater than 0",
+				req.Name, req.Resolution.Width, req.Resolution.Height,
+			)
+	}
+	return optionsCommandResize, nil
+}
```
Comment (randhid) [inline robot/web/stream/server.go:781]: nit: put this above where it's used.

Comment (seanavery) [reply robot/web/stream/server.go]: thanks, moved.

File: robot/web/stream/state/state.go (hunk 10 above)
Comment (seanavery) [reply robot/web/stream/state/state.go:322]: Implemented the `Reset` path for this. Not exactly sure how to handle this from a UI perspective -- may need to update the wire-diagram in the scope doc to include a Reset button.

File: robot/web/stream/state/state.go (hunk 9 above)
Comment (seanavery) [reply robot/web/stream/state/state.go:218]: Thanks for bringing this up -- the Reset path has been implemented to handle getting out of resized mode.

**Review by randhid — APPROVED** [review]

**Pull request merged**